Make ExecuTorch CoreML asset/trash/databse folders configurable.

metascroy opened the pull request:
Summary:
The ExecuTorch CoreML backend uses various folder locations to store compiled assets, to trash old assets, and to store a databse.


With this diff, all of these are configurable with the following compile-time flags:

* EXECUTORCH_COREML_ASSETS_DIRECTORY_PATH (default value is [cache]/executorchcoreml)
* EXECUTORCH_COREML_TRASH_DIRECTORY_PATH (default value is [tmp]/executorchcoreml)
* EXECUTORCH_COREML_DATABASE_DIRECTORY_PATH (default value is [app support]/executorchcoreml)

Differential Revision: D85064722

shoumikhin commented on backends/apple/coreml/runtime/delegate/ETCoreMLStrings.mm:
```diff
@@ -101,39 +101,50 @@ + (NSString *)debugSymbolToHandlesKeyName {
+        static NSString *result = nil;
+        dispatch_once(&onceToken, ^{
+            NSArray<NSString *> *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
+            if (paths.count > 0) {
```
nit: technically, neither of `paths.count > 0` checks are needed

metascroy commented on backends/apple/coreml/runtime/delegate/ETCoreMLStrings.mm:
```diff
@@ -101,39 +101,50 @@ + (NSString *)debugSymbolToHandlesKeyName {
-    });
-    
-    return result;
+    #if defined(EXECUTORCH_COREML_ASSETS_DIRECTORY_PATH)
```
This is a compile time flag for the CoreML backend, not a feature that can be changed between predictions.   If the flag is defined at compile time, users can control what directories are used for model caches.  If not defined, a default folder is used.

cccclai reviewed (approved):
Seems fine to me as it's a requested feature

metascroy committed:
Make ExecuTorch CoreML asset/trash/databse folders configurable. (#15984)

Summary:

The ExecuTorch CoreML backend uses various folder locations to store compiled assets, to trash old assets, and to store a databse.


With this diff, all of these are configurable with the following compile-time flags:

* EXECUTORCH_COREML_ASSETS_DIRECTORY_PATH (default value is [cache]/executorchcoreml)
* EXECUTORCH_COREML_TRASH_DIRECTORY_PATH (default value is [tmp]/executorchcoreml)
* EXECUTORCH_COREML_DATABASE_DIRECTORY_PATH (default value is [app support]/executorchcoreml)

Differential Revision: D85064722
